File: backend/ocr.py
"""
ocr.py — FastAPI router for receipt OCR extraction.
All Gemini calls go through GeminiManager for quota safety.

Endpoints:
    POST /api/ocr-receipt   — Upload a receipt image, get structured expense data
"""

from fastapi import APIRouter, UploadFile, File, HTTPException
from gemini_manager import handle_receipt_upload

router = APIRouter()


@router.post("/api/ocr-receipt")
async def ocr_receipt(receipt: UploadFile = File(...)):
    """
    Upload a receipt image and get AI-extracted expense fields.
    Returns extracted data + Gemini quota stats.
    On quota exhaustion, returns fallback data with a warning.
    """

    # ── Validate file type ────────────────────────────────────────────────
    allowed = {"png", "jpg", "jpeg", "gif", "webp"}
    ext = (receipt.filename or "").rsplit(".", 1)[-1].lower()

    if ext not in allowed:
        raise HTTPException(
            status_code=400,
            detail=f"File type .{ext} not supported. Use JPG or PNG.",
        )

    # ── Read file bytes and pass to GeminiManager ─────────────────────────
    file_bytes = await receipt.read()
    result = handle_receipt_upload(file_bytes)

    return result

File: backend/src/routes/ocr.js
const express = require('express');
const multer = require('multer');
const path = require('path');
const { authenticate, requirePasswordChanged } = require('../middleware/auth');
const { extractReceiptData } = require('../services/ocrService');

const router = express.Router();

// Multer setup for OCR uploads (temporary storage)
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, path.join(__dirname, '..', '..', 'uploads'));
  },
  filename: (req, file, cb) => {
    const ext = path.extname(file.originalname);
    cb(null, `ocr-${Date.now()}-${Math.random().toString(36).substring(7)}${ext}`);
  },
});
const upload = multer({
  storage,
  limits: { fileSize: 10 * 1024 * 1024 }, // 10MB
  fileFilter: (req, file, cb) => {
    const allowedTypes = /jpeg|jpg|png|gif|pdf|webp/;
    const extname = allowedTypes.test(path.extname(file.originalname).toLowerCase());
    const mimetype = allowedTypes.test(file.mimetype);
    if (extname && mimetype) {
      cb(null, true);
    } else {
      cb(new Error('Only image files (JPEG, PNG, GIF, WebP) and PDFs are allowed.'));
    }
  },
});

/**
 * POST /api/ocr/receipt — Upload a receipt image for AI analysis
 * Returns extracted expense data (amount, currency, date, category, vendor, description)
 */
router.post('/ocr/receipt', authenticate, requirePasswordChanged, upload.single('receipt'), async (req, res, next) => {
  try {
    if (!req.file) {
      return res.status(400).json({ error: 'No receipt file uploaded.' });
    }

    console.log(`[OCR] Processing receipt: ${req.file.filename} (${(req.file.size / 1024).toFixed(1)}KB)`);

    const result = await extractReceiptData(req.file.path);

    // Return the uploaded file path so the frontend can attach it to the expense later
    result.receipt_path = `/uploads/${req.file.filename}`;

    res.json(result);
  } catch (err) {
    next(err);
  }
});

module.exports = router;
